rfd/0148-pinned-unified-resources.md

```diff
@@ -0,0 +1,126 @@
+---
+authors: Michael Myers (michael.myers@goteleport.com)
+state: draft
+---
+
+# RFD 0148 - Pinned Unified Resources in the web UI
+
+## Required approvers
+
+TODO
+
+## What
+
+This RFD discusses the method of "pinning" resources in the web UI. Pinning will allow users to have 
+specific resources in a tab that they want to have easier access to. Pinning is analogous to "favoriting"
+
+![unnamed](https://github.com/gravitational/teleport/assets/5201977/affe68b9-323f-4aa0-948c-9d8fb53f8c01)
+![unnamed-1](https://github.com/gravitational/teleport/assets/5201977/1f9c5915-4cde-478c-b788-cd49b04edcd3)
+
+
+
+
+## Why
+
+Generally, most users access some resources much more than others. We reduced friction of search/discovery by
+adding in the unified resource view. Pinning takes this view one step further by allowing them to keep their
+favorite resources always within a click or two away without needing to search and filter for the same resource
+every day. 
+
+## Details
+
+### User Preferences
+
+Pinned resources will be stored on the `UserPrefernces` object as an array of resource IDs (that match the id used in our unified resources caches, ex: `theservername/node`, `grafana/app`, etc). 
+
+
+```diff
+type UserPreferencesResponse struct {
+	Assist AssistUserPreferencesResponse `json:"assist"`
+	Theme userpreferencesv1.Theme `json:"theme"`
+	Onboard OnboardUserPreferencesResponse `json:"onboard"`
++   PinnedResources []string `json:"pinnedResources"`
+}
+```
+
+Defined in protobuf as below:
+```protobuf
+// PinnedResourcesUserPreferences is a collection of resource IDs that will be
+// displayed in the user's pinned resources tab in the Web UI
+message PinnedResourcesUserPreferences {
+	// resource_ids is a list of resource ids
+	repeated string resource_ids = 1;
+}
+```
+Currently, user preferences are only access via the root auth server. This makes sense for things like theme where
+it is expected that the user would want the same theme across all clusters. However, with pinned resources,
+we would want a separate list per cluster. Instead of creating a new mechanism to store pinned resources, we can
+reuse the current user preferences method but update/fetch pinned resources per cluster instead. This would require two 
+new endpoints in the apiserver `pinnedResourcesUpsert` and `pinnedResourcesGet`. 
```

Comment:
Why do we need a new endpoint to update pinned resources? Wouldn't we just use the existing endpoints to update user preferences?

Reply:
> Good question. The difference is, if we go with per-cluster, we will have to have some way to say "this is the overall preferences" and "this is the cluster's preferences".
> I think both can make sense where there are some user settings that you'd expect to be global, like theme, assist settings and what not, and then storing "cluster specific" settings in their respective cluster.
> If thats the case, we can use the same mechanism of user preferences, but just change which auth server we reference. currently, the endpoints now are root only. Instead of changing that (and maybe my proposed endpoints were poorly named) we could have like getClusterUserPreferences instead. Maybe we don't need an endpoint and can just have part of the request specify the cluster to pull from.